Added include and exclude functionality to history component

wokar opened the pull request:
**Description:**
Added `include` and `exclude` configuration to the history component.

Without any `include` or `exclude` configuration the history displays graphs for every entity (well that's not exactly true - for instance `hidden` entities or `scenes` are never shown) on a given date. If you are only interested in some of the entities you several options:
- define domains and entities to `exclude` (aka. blacklist) 
- define domains and entities to display by using the `include` configuration (aka. whitelist).
- use the `include` list to define the domains/entities to display, and exclude some of them with in the `exclude` list. This makes sense if you for instance include the sensor domain, but want to exclude some specific sensor like in the example below. 

**Related issue (if applicable):** fixes #3311

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#1094

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
# Example configuration.yaml entry without any filters
history:

# Example configuration.yaml entry with exclude
history:
  exclude:
    domains:
      - automation
      - sun
      - weblink
      - group
      - updater
    entities:
      - sensor.last_boot
      - sensor.date

# Example configuration.yaml entry with include and exclude
history:
  include:
    domains:
      - sensor
      - switch
      - media_player
  exclude:
    entities:
     - sensor.last_boot
     - sensor.date
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code does not interact with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] Tests have been added to verify that the new code works.

wokar committed:
added include and exclude functionality to history component

wokar committed:
fixed summary lines in test method doc.

arsaboo commented:
Excluding entities from history will also prevent the cards to be loaded when we click on the entity on the front screen? I am guessing yes, but just to make sure.

wokar commented:
No - this PR just deals with the history component. The details view of an entity will just work as before.

wokar committed:
cleanup of query filter creation

arsaboo commented:
@wokar Which is even better :100:

balloob commented on homeassistant/components/history.py:
```diff
@@ -96,23 +118,23 @@ def get_states(utc_point_in_time, entity_ids=None, run=None):
-            states.entity_id.in_(entity_ids))
+        (states.created < utc_point_in_time) &
+        (~states.domain.in_(IGNORE_DOMAINS)))
+    most_recent_state_ids = _set_filters_in_query(
```
Why does this need filtering? If this is to filter entity_ids, those should have been filtered before we got into `get_states` ?

wokar replied:
The get `states_to_json()`method uses the `get_states()` to get the state at start time. If we do not filter here all the filtered entities will show up again in the history, but the only show the start value (a lot of flat lines and bars with one value for the whole day).

balloob commented on homeassistant/components/history.py:
```diff
@@ -140,16 +162,15 @@ def states_to_json(states, start_time, entity_id):
     """Setup the history hooks."""
-    hass.wsgi.register_view(Last5StatesView)
-    hass.wsgi.register_view(HistoryPeriodView)
+    hass.wsgi.register_view(Last5StatesView(hass, config))
```
This one doesn't seem to need the config

balloob commented on homeassistant/components/history.py:
```diff
@@ -185,8 +216,77 @@ def get(self, request, datetime=None):
-        return self.json(
-            get_significant_states(start_time, end_time, entity_id).values())
+        return self.json(get_significant_states(
+            start_time, end_time, entity_id, Filters(self.config)).values())
```
Let's cache the `Filters` object. If you create the filters in `setup` you don't even have to store the `config`.

balloob commented on homeassistant/components/history.py:
```diff
@@ -7,16 +7,37 @@
+            vol.Optional(CONF_DOMAINS, default=[]): cv.ensure_list
+        }),
+        CONF_INCLUDE: vol.Schema({
+            vol.Optional(CONF_ENTITIES, default=[]): cv.ensure_list,
```
If you default it to an empty list, you are unable to differentiate between not getting a value (`None`) or a user wanting to filter out everything (`[]`).

balloob replied:
You probably want to not have any default values for the includes.

balloob commented on homeassistant/components/history.py:
```diff
@@ -185,8 +216,77 @@ def get(self, request, datetime=None):
+            self.included_domains = include[CONF_DOMAINS]
+
+
+def _set_filters_in_query(query, states, entity_ids, filters):
```
Would it make sense to move this to make this an instance method `apply` on the `Filters` class ?

balloob replied:
Also, passing in `states` is confusing (esp the name, since we also have lists/generators of states). I would prefer to use `states = recorder.get_model('States')` inside the method.

balloob commented on homeassistant/components/history.py:
```diff
@@ -185,8 +216,77 @@ def get(self, request, datetime=None):
+class Filters(object):
+    """Container for the configured include and exclude filters."""
+
+    def __init__(self, config):
```
I would prefer the `Filters` class to not be aware of the shape of the config. That way we could potentially re-use this class.

wokar committed:
o improved config validation
o move move IGNORE_DOMAINS to Filter.apply()
o removed config from Last5StatesView
o Filters instance is now created on setup
o config values are processed in setup and set to the Filters instance
o function _set_filters_in_query() moved to Filters class and renamed to apply()

wokar committed:
added unittests for more include/exclude filter combinations

wokar committed:
make pylint happy

balloob commented:
Awesome 🐬
